src/core/unit.c

```diff
@@ -632,8 +670,11 @@ void unit_free(Unit *u) {
 
         unit_free_requires_mounts_for(u);
 
-        SET_FOREACH(t, u->names, i)
-                hashmap_remove_value(u->manager->units, t, u);
+        if (u->names)
+                SET_FOREACH(t, u->names, i)
+                        hashmap_remove_value(u->manager->units, t, u);
+        else if (u->id)
+                hashmap_remove_value(u->manager->units, u->id, u);
```

Comment:
cast to void?

Reply:
None of the other calls to hashmap_remove have the case. I think there is no particular reason to use it here, it seems clear that this may fail.